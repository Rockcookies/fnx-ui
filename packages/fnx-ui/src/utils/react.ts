import { Children, ReactElement } from 'react';
import { isFragment } from 'react-is';

export const toElementArray = (
	children: React.ReactNode,
	option?: { keepEmpty?: boolean },
): ReactElement[] => {
	const keepEmpty = option && option.keepEmpty;
	let ret: React.ReactElement[] = [];

	Children.forEach(children, (child: any) => {
		if ((child === undefined || child === null) && !keepEmpty) {
			return;
		}

		if (Array.isArray(child)) {
			ret = ret.concat(toElementArray(child));
		} else if (isFragment(child) && child.props) {
			ret = ret.concat(toElementArray(child.props.children, option));
		} else {
			ret.push(child);
		}
	});

	return ret;
};
