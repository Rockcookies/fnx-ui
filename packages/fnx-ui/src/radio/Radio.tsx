import { forwardRef, useContext } from 'react';
import useCheckboxRender from '../checkbox/hooks/use-checkbox-render';
import useControlledState from '../hooks/use-controlled-state';
import useDefaults from '../hooks/use-defaults';
import useProps from '../hooks/use-props';
import { createBEM } from '../utils/namespace';
import { RadioGroupContext } from './context';
import { RadioProps } from './interface';

const NS = 'fnx-radio';
const bem = createBEM(NS);

type RadioRequiredProps = Required<
	Pick<RadioProps, 'defaultChecked' | 'skipGroup'>
>;

const DEFAULT_PROPS: RadioRequiredProps = {
	defaultChecked: false,
	skipGroup: false,
};

const Radio = forwardRef<HTMLDivElement, RadioProps>((_props, ref) => {
	const [
		{ defaultChecked, skipGroup },
		{
			checked: _checked,
			value: radioValue,
			onChange: _onChange,
			...resetProps
		},
	] = useProps<RadioRequiredProps, RadioProps>(DEFAULT_PROPS, _props);

	const groupContext = useContext(RadioGroupContext);

	const { value: checked, onChangeRef } = useControlledState<boolean>({
		value:
			groupContext.onChange != null && !skipGroup
				? groupContext.value === radioValue
				: _checked,
		defaultValue: skipGroup ? false : defaultChecked,
		onChange: _onChange,
	});

	const disabled = useDefaults<boolean>(
		false,
		resetProps.disabled,
		groupContext.disabled,
	);

	const render = useCheckboxRender(
		{
			role: 'radio',
			...resetProps,
			onClick: (e) => {
				if (!disabled) {
					if (
						!skipGroup &&
						groupContext.onChange != null &&
						radioValue != null
					) {
						if (groupContext.value !== radioValue) {
							groupContext.onChange(radioValue);
						}
					} else {
						!checked && onChangeRef.current(true);
					}
				}

				resetProps.onClick && resetProps.onClick(e);
			},
		},
		groupContext,
	);

	return render({
		bem,
		checked,
		ref: ref as any,
	});
});

Radio.displayName = 'Radio';

export default Radio;
