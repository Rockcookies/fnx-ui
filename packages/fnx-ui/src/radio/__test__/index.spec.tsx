import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import TestsDOM from '../../../test/dom';
import Radio from '../';

describe('<Radio/>', () => {
	it('disabled prop', async () => {
		const { container } = render(<Radio disabled={true}>box1</Radio>);
		const icon = TestsDOM.mustQuerySelector(container, '.fnx-radio__icon');

		fireEvent.click(icon);
		expect(container.querySelector('.fnx-radio__icon--checked')).toBeNull();
	});

	it('labelDisabled prop', async () => {
		const { container, rerender } = render(
			<Radio labelDisabled={true}>box1</Radio>,
		);
		const icon = TestsDOM.mustQuerySelector(container, '.fnx-radio__icon');
		const label = TestsDOM.mustQuerySelector(
			container,
			'.fnx-radio__label',
		);

		fireEvent.click(label);
		expect(icon.classList.contains('fnx-radio__icon--checked')).toBeFalsy();

		rerender(<Radio labelDisabled={false}>box1</Radio>);

		fireEvent.click(label);
		expect(
			icon.classList.contains('fnx-radio__icon--checked'),
		).toBeTruthy();
	});

	it('iconSize、iconPosition、iconShape、iconCheckedColor', async () => {
		const { container } = render(
			<Radio
				defaultChecked={true}
				iconSize={20}
				iconPosition="right"
				iconShape="square"
				iconCheckedColor="red"
			>
				box1
			</Radio>,
		);
		const icon = TestsDOM.mustQuerySelector(container, '.fnx-radio__icon');

		const style = getComputedStyle(icon);

		expect(style.fontSize).toEqual('20px');
		expect(style.backgroundColor).toEqual('red');
		expect(style.borderColor).toEqual('red');
		expect(icon.classList.contains('fnx-radio__icon--square')).toBeTruthy();
	});

	it('icon & checkedIcon prop', () => {
		const { container } = render(
			<Radio
				iconShape="plain"
				checkedIcon={<span>checked</span>}
				icon={<span>no-check</span>}
			>
				box1
			</Radio>,
		);

		expect(container).toMatchSnapshot();
	});

	it('Radio onChange', () => {
		const onChange = jest.fn();
		const { container } = render(<Radio onChange={onChange}>box1</Radio>);

		const icon = TestsDOM.mustQuerySelector(container, '.fnx-radio__icon');

		fireEvent.click(icon);
		expect(onChange).toHaveBeenCalledWith(true);

		fireEvent.click(icon);
		expect(onChange).toHaveBeenCalledTimes(1);
	});

	it('skipGroup prop', () => {
		const { container } = render(
			<Radio.Group defaultValue="1">
				<Radio skipGroup={true} value="1">
					box1
				</Radio>
				<Radio value="2">box2</Radio>
				<Radio value="3">box3</Radio>
			</Radio.Group>,
		);

		const icons = container.querySelectorAll('.fnx-radio__icon');
		expect(
			icons[0].classList.contains('fnx-radio__icon--checked'),
		).not.toBeTruthy();

		fireEvent.click(icons[1]);
		expect(
			icons[1].classList.contains('fnx-radio__icon--checked'),
		).toBeTruthy();
	});

	it('RadioGroup onChange', () => {
		const onChange = jest.fn();
		const { container } = render(
			<Radio.Group onChange={onChange}>
				<Radio skipGroup={true} value="1">
					box1
				</Radio>
				<Radio value="2">box2</Radio>
				<Radio value="3">box3</Radio>
			</Radio.Group>,
		);

		const icons = container.querySelectorAll('.fnx-radio__icon');
		fireEvent.click(icons[0]);
		expect(onChange).not.toHaveBeenCalled();

		fireEvent.click(icons[1]);
		expect(onChange).toHaveBeenCalledWith('2');

		fireEvent.click(icons[2]);
		expect(onChange).toHaveBeenCalledWith('3');
	});
});
