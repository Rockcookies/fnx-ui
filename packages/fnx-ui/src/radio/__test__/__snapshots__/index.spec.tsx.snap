// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio/> icon & checkedIcon prop 1`] = `
<div>
  <div
    aria-checked="false"
    class="fnx-radio"
    role="radio"
    tabindex="0"
  >
    <span
      class="fnx-radio__icon"
    >
      <span>
        no-check
      </span>
    </span>
    <span
      class="fnx-radio__label"
    >
      box1
    </span>
  </div>
</div>
`;
